### DISCUSSION TITLE
Type constraint mismatch when reraising a caught exception in a task CE

### DISCUSSION THREAD

**Discussion by adam-c-anderson** [dotnet/fsharp#14482] [Q&A] [answered]
I have some code somewhat like the following: (actual types obfuscated)

```fs
// unit -> Task<MyClass option>
member this.DoThing ()=
    task {
        try
            let! result = something.doAsync ()
            return result // I know, this could've just been return!, please overlook
        with
        | ex ->
            logger.LogError(ex, "Error doing the thing")
            raise ex // I know this doesn't preserve the stack trace, please overlook
            return None // Won't compile without this line
    }
```

I am very confused by the compile-time behavior of this. Why must the `with` expression include a `return` after calling `raise`? If the `return None` is omitted, I get a compilation error like this:

```
  [FS0193] Type constraint mismatch. The type 
    'TaskCode<MyClass option,unit>'    
is not compatible with type
    'TaskCode<MyClass option,MyClass option>'
```

I found the docs that say `TaskCode<,>` is an abbreviation for resumable code, though that doesn't bring me closer to understanding. I understand why the `try` and `with` expressions must return the same type, but shouldn't that requirement be dismissed by (re)raising an exception? Would `return None` even get executed even though the compiler requires it?

**Comment by baronfel** [comment, marked as answer]
All you should have to do here is `return` the `raise ex` expression. `raise` (like `throw`, `failwith` and others), returns a generic type so that it can be used in situations like this, you just always have to `return` something out of a CE context.

**Reply by adam-c-anderson** [reply]
That certainly seems to have worked. Thanks!
